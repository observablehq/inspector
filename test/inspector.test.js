/* eslint-env jest */
const { Inspector } = require("../dist/notebook-inspector.umd.js");

describe("Inspector", () => {
  let inspector, elem;
  beforeEach(() => {
    elem = document.createElement("div");
    inspector = new Inspector(elem);
  });

  test("initial state", () => {
    expect(elem).toMatchSnapshot();
  });

  test(".pending()", () => {
    inspector.pending();
    expect(elem).toMatchSnapshot();
  });

  test(".fulfilled(element)", () => {
    const span = document.createElement("span");
    span.textContent = "Surprise!";
    inspector.fulfilled(span);
    expect(elem).toMatchSnapshot();
  });

  test(".fulfilled(value)", () => {
    inspector.fulfilled([1, 2, 3]);
    expect(elem).toMatchSnapshot();
    elem.querySelector("a").dispatchEvent(new MouseEvent("mouseup"));
    expect(elem).toMatchSnapshot();
  });

  test(".rejected(Error)", () => {
    inspector.rejected(new Error("Danger!"));
    expect(elem).toMatchSnapshot();
  });
});

describe("into", () => {
  test("into('invalid id')", () => {
    expect(() => {
      Inspector.into("unknown-id");
    }).toThrow();
  });

  test("into(div)", () => {
    const container = document.createElement("div");
    const inspector = Inspector.into(container)();
    inspector.fulfilled(42);
    expect(container).toMatchSnapshot();
  });

  test("formats a string with template syntax if it has multiple newlines", () => {
    const container = document.createElement("div");
    const inspector = Inspector.into(container)();
    inspector.fulfilled(Array.from({ length: 10 }, () => "hi").join("\n"));
    expect(container).toMatchSnapshot();
  });

  test("formats a string with JSON syntax if it doesnâ€™t have many newlines", () => {
    const container = document.createElement("div");
    const inspector = Inspector.into(container)();
    inspector.fulfilled(Array.from({ length: 10 }, () => "hi").join(" "));
    expect(container).toMatchSnapshot();
  });

  test("truncates a string with > 20 newlines", () => {
    const container = document.createElement("div");
    const inspector = Inspector.into(container)();
    inspector.fulfilled(
      Array.from({ length: 30 }, () => "hi").join("\n"),
      "myString"
    );
    expect(container).toMatchSnapshot();
    container
      .querySelector(".observablehq--string-expand")
      .dispatchEvent(new MouseEvent("mouseup"));
    expect(container).toMatchSnapshot();
  });

  test("truncates a string with > 20 newlines", () => {
    const container = document.createElement("div");
    const inspector = Inspector.into(container)();
    inspector.fulfilled(
      Array.from({ length: 21 }, () => "hi").join("\n"),
      "myString"
    );
    expect(container).toMatchSnapshot();
  });
});
