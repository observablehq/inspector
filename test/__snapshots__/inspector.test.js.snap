// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inspector .fulfilled(element) 1`] = `
<div
  class="observablehq"
>
  <span>
    Surprise!
  </span>
</div>
`;

exports[`Inspector .fulfilled(value) 1`] = `
<div
  class="observablehq"
>
  <span
    class="observablehq--collapsed observablehq--inspect"
  >
    <a>
      <svg
        class="observablehq--caret"
        height="8"
        width="8"
      >
        
    
        <path
          d="M7 4L1 8V0z"
          fill="currentColor"
        />
        
  
      </svg>
      Array(3) [
    </a>
    <span
      class="observablehq--number"
    >
      1
    </span>
    , 
    <span
      class="observablehq--number"
    >
      2
    </span>
    , 
    <span
      class="observablehq--number"
    >
      3
    </span>
    ]
  </span>
</div>
`;

exports[`Inspector .fulfilled(value) 2`] = `
<div
  class="observablehq"
>
  <span
    class="observablehq--expanded observablehq--inspect"
  >
    <a>
      <svg
        class="observablehq--caret"
        height="8"
        width="8"
      >
        
    
        <path
          d="M4 7L0 1h8z"
          fill="currentColor"
        />
        
  
      </svg>
      Array(3) [
    </a>
    <div
      class="observablehq--field"
    >
      <span
        class="observablehq--index"
      >
          0
      </span>
      : 
      <span
        class="observablehq--number"
      >
        1
      </span>
    </div>
    <div
      class="observablehq--field"
    >
      <span
        class="observablehq--index"
      >
          1
      </span>
      : 
      <span
        class="observablehq--number"
      >
        2
      </span>
    </div>
    <div
      class="observablehq--field"
    >
      <span
        class="observablehq--index"
      >
          2
      </span>
      : 
      <span
        class="observablehq--number"
      >
        3
      </span>
    </div>
    ]
  </span>
</div>
`;

exports[`Inspector .pending() 1`] = `
<div
  class="observablehq observablehq--running"
/>
`;

exports[`Inspector .rejected(Error) 1`] = `
<div
  class="observablehq observablehq--error"
>
  <span
    class="observablehq--inspect"
  >
    Error: Danger!
  </span>
</div>
`;

exports[`Inspector initial state 1`] = `
<div
  class="observablehq"
/>
`;

exports[`into formats a string with JSON syntax if it doesnâ€™t have many newlines 1`] = `
<div>
  <div
    class="observablehq"
  >
    <span
      class="observablehq--string observablehq--inspect"
    >
      "hi hi hi hi hi hi hi hi hi hi"
    </span>
  </div>
</div>
`;

exports[`into formats a string with template syntax if it has multiple newlines 1`] = `
<div>
  <div
    class="observablehq"
  >
    <span
      class="observablehq--string observablehq--inspect"
    >
      \`hi
hi
hi
hi
hi
hi
hi
hi
hi
hi\`
    </span>
  </div>
</div>
`;

exports[`into into(div) 1`] = `
<div>
  <div
    class="observablehq"
  >
    <span
      class="observablehq--number observablehq--inspect"
    >
      42
    </span>
  </div>
</div>
`;

exports[`into truncates a string with > 20 newlines 1`] = `
<div>
  <div
    class="observablehq"
  >
    <div
      class="observablehq--string observablehq--inspect"
    >
      \`hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
      <span
        class="observablehq--string-expand"
      >
        Show 10 truncated lines
      </span>
    </div>
  </div>
</div>
`;

exports[`into truncates a string with > 20 newlines 2`] = `
<div>
  <div
    class="observablehq"
  >
    <span
      class="observablehq--string observablehq--expanded observablehq--inspect"
    >
      \`hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi\`
    </span>
  </div>
</div>
`;

exports[`into truncates a string with > 20 newlines 3`] = `
<div>
  <div
    class="observablehq"
  >
    <div
      class="observablehq--string observablehq--inspect"
    >
      \`hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
hi
      <span
        class="observablehq--string-expand"
      >
        Show 1 truncated line
      </span>
    </div>
  </div>
</div>
`;
